core-primitives/settings/src/lib.rs

```diff
@@ -53,6 +53,7 @@ pub mod files {
 	pub static RA_API_KEY_FILE: &str = "key.txt";
 
 	pub const SPID_MIN_LENGTH: usize = 32;
+	pub const STATE_SNAPSHOTS_CACHE_SIZE: usize = 40;
```

Comment:
The number of state snapshots we keep. 40 = the number of sidechain blocks (300 ms block time) during a parentchain block production cycle (12s block time).

Reply:
You are probably right. I thought only about finalizing sidechain blocks, but not about finalizing the parentchain block too, that finalizes the sidechain blocks. So if that takes about 3 parentchain blocks to finalize, we're at 120 state snapshots (which will potentially take up a lot of space 😅).

The idea to fetch the 'correct' state from a peer is interesting, we haven't considered that yet in our fork mitigation design. We planned on rolling back the state and then peer-fetch the blocks.

Reply:
Agreed. I also believe interval snapshots should suffice. But regarding the performance issue, couldn't the untrusted side take care of handling the snapshots? The trusted side only needs to know where to find the snapshots.

Reply:
Well, we need to snapshot each state, because we cannot know to which state we need to roll back to. Writing to disk or not is a performance optimization that we planned for the caching mechanism. IMO that is out of scope of this PR, introducing a caching mechanism would blow it completely out of proportion. But yes, it is something we agreed that we need in the future.

Reply:
> I also worry about performance if we write the whole state to disc with every sidechain block.

That's what we do so far already, this PR changes nothing about that behavior. What we have now, are more state files. But the amount of disk I/O remains the same.

Reply:
> Agreed. I also believe interval snapshots should suffice. But regarding the performance issue, couldn't the untrusted side take care of handling the snapshots? The trusted side only needs to know where to find the snapshots.

Storing only every n-th snapshot could be an option, yes. We would need to think about how reverting in that case works. This would reduce the amount of files and disk space we use.
But I think I don't understand the point of the untrusted side - what benefit would we have from doing any of this on the untrusted side? 

In any case: Can we agree that we see 2 enhancements here, that we should explore and discuss (but not in the context of this PR and issue) ?:
1. In-memory cache to reduce amount of file I/O necessary
2. Store snapshots only at intervals, not for every state update, reducing the amount of stored files, saving disk space.

Reply:
Agreed. Can you create an (enhancement? or maybe a new label "performance"?) issue für this?

Reply:
yes: #706

Reply:
Ok, you are right. we are already writing to state every time. :P I forgot that, sorry!

I can live very well with the enhancement issue, and merge this PR as is. 👍